### PR TITLE
Send OBSTACLE_DISTANCE from AP_Proximity backends

### PR DESCRIPTION
**Pull request by peterbarker**
... also fix RPiLIDAR

### CODE REVIEW

**Comment by peterbarker** [comment]
I've now tested this on a real-world RPLidar2 and it works well.

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 1]
```diff
@@ -4312,6 +4313,17 @@ void GCS_MAVLINK::send_sys_status()
         errors4); // errors4
 }
 
+
```
Comment (rmackay9) [inline libraries/GCS_MAVLink/GCS_Common.cpp:4316]: some extra lines we can probably remove here.

Comment (peterbarker) [reply libraries/GCS_MAVLink/GCS_Common.cpp]: Fixed.

File: libraries/AP_Proximity/AP_Proximity.cpp [hunk 2]
```diff
@@ -175,6 +177,22 @@ const AP_Param::GroupInfo AP_Proximity::var_info[] = {
     AP_GROUPINFO("2_YAW_CORR", 18, AP_Proximity, _yaw_correction[1], 0),
 #endif
 
+    // @Param: _OPTIONS
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity.cpp:180]: as discussed let's remove these extra parameters.  It's just getting too confusing for users to figure out how to set things up.  Once @meee1 has completed MP's obstacle distance viewer (and we've confirmed it works) then we should stop sending the distance-sensor messages from the proximity library.

Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity.cpp]: Done.

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp [hunk 3]
```diff
@@ -31,6 +33,10 @@ AP_Proximity_Backend::AP_Proximity_Backend(AP_Proximity &_frontend, AP_Proximity
 {
     // initialise sector edge vector used for building the boundary fence
     init_boundary();
+
+    obstacle_distance.increment = 360/PROXIMITY_NUM_SECTORS;
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_Backend.cpp:37]: let's always set the increment_f as well just to be nice to the consumer.

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 4]
```diff
@@ -63,6 +63,17 @@
 
 extern const AP_HAL::HAL& hal;
 
+AP_Proximity_RPLidarA2::AP_Proximity_RPLidarA2(AP_Proximity &_frontend, AP_Proximity::Proximity_State &_state) :
+    AP_Proximity_Backend_Serial(_frontend, _state)
+{
+    // we can fill in all 72 distances:
+    obstacle_distance.increment = 5;
```
Comment (rmackay9) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:70]: let's fille in increment_f here as well I think..

**Comment by rmackay9** [comment]
Txs for this.  One key requirement is that backends that receive the obstacle-distance messages also send the obstacle-distance message.  This will allow us to turn off the high-rate obstacle-distance message sent directly from the sensor and replace it with a lower rate message from AP.

File: libraries/AP_Proximity/AP_Proximity_Backend.cpp (hunk 3 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_Backend.cpp:37]: Done.

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 4 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:70]: Done.

**Comment by peterbarker** [comment]
@rmackay9 as requested, I've handled the case where the sensor stops feeding us data or in some other ways goes unhealthy.

It starts to send all distances through as "invalid" - in MAVProxy the arcs representing obstacles disappear.  This was tested by hacking up the simulated RPi sensor:

```
--- a/libraries/SITL/SIM_PS_RPLidarA2.cpp
+++ b/libraries/SITL/SIM_PS_RPLidarA2.cpp
@@ -206,6 +206,9 @@ void PS_RPLidarA2::update_output(const Location &location)
 
 void PS_RPLidarA2::update(const Location &location)
 {
+    if (AP_HAL::millis() > 30000) {
+        return;
+    }
     update_input();
     update_output(location);
 }

```

**Comment by peterbarker** [comment]
@rmackay9 just reread your last comment here.  The *proximity* library sends the obstacle distance message - it's backend-agnostic.  That means if you have only the cardinal points - like the SITL backend - then you end up sending an `OBSTACLE_DISTANCE` with just 8 points in it.  This has been tested in SITL.

**Comment by rmackay9** [comment]
On today's dev call the major point of concern was around the increased telemetry bandwidth that could be consumed.

- Tridge suggested a parameter to allow the user to select whether they wanted DISTANCE_SENSOR or OBSTACLE_DISTANCE messages to be sent.
- Peter suggested that we could leave the OBSTACLE_DISTANCE message out of stream3 and instead require the GCSs to request it using message interval
- Randy suggested that we could add some better checks to decide whether to send distance-sensor or obstacle-distance message based upon the type of proximity sensor being used.  For example if a 360 lidar is used (SF40c, RPLidar, etc) we could send the obstacle-distance message, if only 1D lidar are being used then send distance-sensor messages.

**Pull request closed without merging**